src/v/cluster/cluster_discovery.cc

```diff
@@ -31,12 +31,13 @@ namespace cluster {
 
 cluster_discovery::cluster_discovery(
   const model::node_uuid& node_uuid,
-  storage::kvstore& kvstore,
+  storage::api& storage,
   ss::abort_source& as)
   : _node_uuid(node_uuid)
   , _join_retry_jitter(config::shard_local_cfg().join_retry_timeout_ms())
   , _join_timeout(std::chrono::seconds(2))
-  , _kvstore(kvstore)
+  , _stored_cluster_uuid(storage.get_cluster_uuid())
```

Comment:
nit: I think it's more important for us to cache whether or not we are a cluster founder, rather than whether we have a local cluster UUID. All the decisions we are making depend on the former, not the latter.